File: tests/src/test/javascript/msg/MessageCapabilitiesTest.js
/**
 * Copyright (c) 2013-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Message capabilities unit tests.
 * 
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
describe("MessageCapabilities", function() {
    var MslEncoderFormat = require('msl-core/io/MslEncoderFormat.js');
    var MslConstants = require('msl-core/MslConstants.js');
    var EntityAuthenticationScheme = require('msl-core/entityauth/EntityAuthenticationScheme.js');
    var MessageCapabilities = require('msl-core/msg/MessageCapabilities.js');
    var Arrays = require('msl-core/util/Arrays.js');

    var MslTestConstants = require('msl-tests/MslTestConstants.js');
    var MockMslContext = require('msl-tests/util/MockMslContext.js');
    var MslTestUtils = require('msl-tests/util/MslTestUtils.js');
    
    /** MSL encoder format. */
    var ENCODER_FORMAT = MslEncoderFormat.JSON;
    
    /** Key compression algorithms. */
    var KEY_COMPRESSION_ALGOS = "compressionalgos";
    /** Key encoder formats. */
    var KEY_ENCODER_FORMATS = "encoderformats";
    
    // Shortcuts
    var CompressionAlgorithm = MslConstants.CompressionAlgorithm;
    
    var ALGOS = [ CompressionAlgorithm.GZIP, CompressionAlgorithm.LZW ];
    var LANGUAGES = [ "en-US", "es" ];
    var FORMATS = [ MslEncoderFormat.JSON ];
    
    /** MSL encoder factory. */
    var encoder;
    
    var initialized = false;
    beforeEach(function() {
        if (!initialized) {
            var ctx;
            runs(function() {
                MockMslContext.create(EntityAuthenticationScheme.PSK, false, {
                    result: function(x) { ctx = x; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); },
                });
            });
            waitsFor(function() { return ctx; }, "ctx", MslTestConstants.TIMEOUT_CTX);
            
            runs(function() {
                encoder = ctx.getMslEncoderFactory();
                initialized = true;
            });
        }
    });
    
    it("ctors", function() {
        var caps = new MessageCapabilities(ALGOS, LANGUAGES, FORMATS);
        expect(caps.compressionAlgorithms).toEqual(ALGOS);
        expect(caps.languages).toEqual(LANGUAGES);
        expect(caps.encoderFormats).toEqual(FORMATS);
        
        var encode;
        runs(function() {
            caps.toMslEncoding(encoder, ENCODER_FORMAT, {
                result: function(x) { encode = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); },
            });
        });
        waitsFor(function() { return encode; }, "encode", MslTestConstants.TIMEOUT);
        
        var moCaps, moEncode;
        runs(function() {
            expect(encode).not.toBeNull();

            moCaps = MessageCapabilities.parse(encoder.parseObject(encode));
            expect(moCaps.compressionAlgorithms).toEqual(caps.compressionAlgorithms);
            expect(moCaps.languages).toEqual(caps.languages);
            expect(moCaps.encoderFormats).toEqual(caps.encoderFormats);
            moCaps.toMslEncoding(encoder, ENCODER_FORMAT, {
                result: function(x) { moEncode = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); },
            });
        });
        waitsFor(function() { return moEncode; }, "moEncode", MslTestConstants.TIMEOUT);
        
        runs(function() {
            expect(moEncode).not.toBeNull();
            // This test will not always pass since set data is unordered.
            //expect(moEncode).toEqual(encode);
            var mo2Caps = MessageCapabilities.parse(encoder.parseObject(moEncode));
            expect(mo2Caps).toEqual(moCaps);
        });
    });
    
    it("ctors with null algorithms", function() {
        var caps = new MessageCapabilities(null, LANGUAGES, FORMATS);
        var algos = caps.compressionAlgorithms;
        expect(algos).not.toBeNull();
        expect(algos.length).toEqual(0);
        expect(caps.languages).toEqual(LANGUAGES);
        expect(caps.encoderFormats).toEqual(FORMATS);
        
        var encode;
        runs(function() {
            caps.toMslEncoding(encoder, ENCODER_FORMAT, {
                result: function(x) { encode = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); },
            });
        });
        waitsFor(function() { return encode; }, "encode", MslTestConstants.TIMEOUT);
        
        var moCaps, moEncode;
        runs(function() {
            expect(encode).not.toBeNull();
            
            moCaps = MessageCapabilities.parse(encoder.parseObject(encode));
            expect(moCaps.compressionAlgorithms).toEqual(caps.compressionAlgorithms);
            expect(moCaps.languages).toEqual(caps.languages);
            expect(moCaps.encoderFormats).toEqual(caps.encoderFormats);
            moCaps.toMslEncoding(encoder, ENCODER_FORMAT, {
                result: function(x) { moEncode = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); },
            });
        });
        waitsFor(function() { return moEncode; }, "moEncode", MslTestConstants.TIMEOUT);
        
        runs(function() {
            expect(moEncode).not.toBeNull();
            // This test will not always pass since set data is unordered.
            //expect(moEncode).toEqual(encode);
            var mo2Caps = MessageCapabilities.parse(encoder.parseObject(moEncode));
            expect(mo2Caps).toEqual(moCaps);
        });
    });
    
    it("unknown compression algorithm", function() {
        var caps = new MessageCapabilities(ALGOS, LANGUAGES, FORMATS);
        var mo;
        runs(function() {
            MslTestUtils.toMslObject(encoder, caps, {
                result: function(x) { mo = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); },
            });
        });
        waitsFor(function() { return mo; }, "mo", MslTestConstants.TIMEOUT);
        
        runs(function() {
            var ma = mo.getMslArray(KEY_COMPRESSION_ALGOS);
            ma.put(-1, "CATZ");
            mo.put(KEY_COMPRESSION_ALGOS, ma);
        
            var moCaps = MessageCapabilities.parse(mo);
            expect(moCaps.compressionAlgorithms).toEqual(caps.compressionAlgorithms);
        });
    });
    
    it("ctors with null languages", function() {
        var caps = new MessageCapabilities(ALGOS, null, FORMATS);
        expect(caps.compressionAlgorithms).toEqual(ALGOS);
        var languages = caps.languages;
        expect(languages).not.toBeNull();
        expect(languages.length).toEqual(0);
        expect(caps.encoderFormats).toEqual(FORMATS);
        
        var encode;
        runs(function() {
            caps.toMslEncoding(encoder, ENCODER_FORMAT, {
                result: function(x) { encode = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); },
            });
        });
        waitsFor(function() { return encode; }, "encode", MslTestConstants.TIMEOUT);
        
        var moCaps, moEncode;
        runs(function() {
            moCaps = MessageCapabilities.parse(encoder.parseObject(encode));
            expect(moCaps.compressionAlgorithms).toEqual(caps.compressionAlgorithms);
            expect(moCaps.languages).toEqual(caps.languages);
            expect(moCaps.encoderFormats).toEqual(caps.encoderFormats);
            moCaps.toMslEncoding(encoder, ENCODER_FORMAT, {
                result: function(x) { moEncode = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); },
            });
        });
        waitsFor(function() { return moEncode; }, "moEncode", MslTestConstants.TIMEOUT);
        
        runs(function() {
            expect(moEncode).not.toBeNull();
            // This test will not always pass since set data is unordered.
            //expect(moEncode).toEqual(encode);
            var mo2Caps = MessageCapabilities.parse(encoder.parseObject(moEncode));
            expect(mo2Caps).toEqual(moCaps);
        });
    });
    
    it("ctors with null encoder formats", function() {
        var caps = new MessageCapabilities(ALGOS, LANGUAGES, null);
        expect(caps.compressionAlgorithms).toEqual(ALGOS);
        expect(caps.languages).toEqual(LANGUAGES);
        var formats = caps.encoderFormats;
        expect(formats).not.toBeNull();
        expect(formats.length).toEqual(0);
        
        var encode;
        runs(function() {
            caps.toMslEncoding(encoder, ENCODER_FORMAT, {
                result: function(x) { encode = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); },
            });
        });
        waitsFor(function() { return encode; }, "encode", MslTestConstants.TIMEOUT);
        
        var moCaps, moEncode;
        runs(function() {
            moCaps = MessageCapabilities.parse(encoder.parseObject(encode));
            expect(moCaps.compressionAlgorithms).toEqual(caps.compressionAlgorithms);
            expect(moCaps.languages).toEqual(caps.languages);
            expect(moCaps.encoderFormats).toEqual(caps.encoderFormats);
            moCaps.toMslEncoding(encoder, ENCODER_FORMAT, {
                result: function(x) { moEncode = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); },
            });
        });
        waitsFor(function() { return moEncode; }, "moEncode", MslTestConstants.TIMEOUT);
        
        runs(function() {
            expect(moEncode).not.toBeNull();
            // This test will not always pass since set data is unordered.
            //expect(moEncode).toEqual(encode);
            var mo2Caps = MessageCapabilities.parse(encoder.parseObject(moEncode));
            expect(mo2Caps).toEqual(moCaps);
        });
    });
    
    it("equals compression algorithm", function() {
        var algosA = Arrays.copyOf(ALGOS);
        var algosB = [];
        
        var capsA = new MessageCapabilities(algosA, LANGUAGES, FORMATS);
        var capsB = new MessageCapabilities(algosB, LANGUAGES, FORMATS);
        var capsA2;
        runs(function() {
            MslTestUtils.toMslObject(encoder, capsA, {
                result: function(mo) {
                    capsA2 = MessageCapabilities.parse(mo);
                },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); },
            });
        });
        waitsFor(function() { return capsA2; }, "capsA2", MslTestConstants.TIMEOUT);
        
        runs(function() {
            expect(capsA.equals(capsA)).toBeTruthy();
            expect(capsA.uniqueKey()).toEqual(capsA.uniqueKey());
            
            expect(capsA.equals(capsB)).toBeFalsy();
            expect(capsB.equals(capsA)).toBeFalsy();
            expect(capsA.uniqueKey()).not.toEqual(capsB.uniqueKey());
            
            expect(capsA.equals(capsA2)).toBeTruthy();
            expect(capsA2.equals(capsA)).toBeTruthy();
            expect(capsA2.uniqueKey()).toEqual(capsA.uniqueKey());
        });
    });
    
    it("equals languages", function() {
        var langsA = [ "en-US" ];
        var langsB = [ "es" ];
        
        var capsA = new MessageCapabilities(ALGOS, langsA, FORMATS);
        var capsB = new MessageCapabilities(ALGOS, langsB, FORMATS);
        var capsA2;
        runs(function() {
            MslTestUtils.toMslObject(encoder, capsA, {
                result: function(mo) {
                    capsA2 = MessageCapabilities.parse(mo);
                },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); },
            });
        });
        waitsFor(function() { return capsA2; }, "capsA2", MslTestConstants.TIMEOUT);
        
        runs(function() {
            expect(capsA.equals(capsA)).toBeTruthy();
            expect(capsA.uniqueKey()).toEqual(capsA.uniqueKey());
            
            expect(capsA.equals(capsB)).toBeFalsy();
            expect(capsB.equals(capsA)).toBeFalsy();
            expect(capsA.uniqueKey()).not.toEqual(capsB.uniqueKey());
            
            expect(capsA.equals(capsA2)).toBeTruthy();
            expect(capsA2.equals(capsA)).toBeTruthy();
            expect(capsA2.uniqueKey()).toEqual(capsA.uniqueKey());
        });
    });
    
    it("equals encoder formats", function() {
        var formatsA = FORMATS;
        var formatsB = [ ];
        
        var capsA = new MessageCapabilities(ALGOS, LANGUAGES, formatsA);
        var capsB = new MessageCapabilities(ALGOS, LANGUAGES, formatsB);
        var capsA2;
        runs(function() {
            MslTestUtils.toMslObject(encoder, capsA, {
                result: function(mo) {
                    capsA2 = MessageCapabilities.parse(mo);
                },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); },
            });
        });
        waitsFor(function() { return capsA2; }, "capsA2", MslTestConstants.TIMEOUT);
        
        runs(function() {
            expect(capsA.equals(capsA)).toBeTruthy();
            expect(capsA.uniqueKey()).toEqual(capsA.uniqueKey());
            
            expect(capsA.equals(capsB)).toBeFalsy();
            expect(capsB.equals(capsA)).toBeFalsy();
            expect(capsA.uniqueKey()).not.toEqual(capsB.uniqueKey());
            
            expect(capsA.equals(capsA2)).toBeTruthy();
            expect(capsA2.equals(capsA)).toBeTruthy();
            expect(capsA2.uniqueKey()).toEqual(capsA.uniqueKey());
        });
    });

    it("intersection with self", function() {
        var capsA = new MessageCapabilities(ALGOS, LANGUAGES, FORMATS);
        var capsB = new MessageCapabilities(ALGOS, LANGUAGES, FORMATS);
        var intersection = MessageCapabilities.intersection(capsA, capsB);
        
        expect(intersection).toEqual(capsA);
        expect(intersection).toEqual(capsB);
    });
    
    if("intersection", function() {
        var gzipOnly = [ CompressionAlgorithm.GZIP ];
        var oneLanguage = [ LANGUAGES[0] ];
        var noFormats = [];
        
        var capsA = new MessageCapabilities(ALGOS, oneLanguage, FORMATS);
        var capsB = new MessageCapabilities(gzipOnly, LANGUAGES, FORMATS);
        var capsC = new MessageCapabilities(ALGOS, LANGUAGES, noFormats);
        var intersectionAB = MessageCapabilities.intersection(capsA, capsB);
        var intersectionBA = MessageCapabilities.intersection(capsB, capsA);
        var intersectionAC = MessageCapabilities.intersection(capsA, capsC);
        var intersectionCA = MessageCapabilities.intersection(capsC, capsA);
        var intersectionBC = MessageCapabilities.intersection(capsB, capsC);
        var intersectionCB = MessageCapabilities.intersection(capsC, capsB);
        
        expect(intersectionAB).toEqual(intersectionBA);
        expect(gzipOnly).toEqual(intersectionAB.compressionAlgorithms);
        expect(Arrays.containEachOther(oneLanguage, intersectionAB.languages)).toBeTruthy();
        expect(FORMATS).toEqual(intersectionAB.encoderFormats);
        
        expect(intersectionAC).toEqual(intersectionCA);
        expect(ALGOS).toEqual(intersectionAC.compressionAlgorithms);
        expect(Arrays.containEachOther(oneLanguage, intersectionAC.languages)).toBeTruthy();
        expect(noFormats).toEqual(intersectionAC.encoderFormats);
        
        expect(intersectionBC).toEqual(intersectionCB);
        expect(gzipOnly).toEqual(intersectionBC.compressionAlgorithms);
        expect(Arrays.containEachOther(LANGUAGES, intersectionBC.languages)).toBeTruthy();
        expect(noFormats).toEqual(intersectionBC.encoderFormats);
    });
    
    it("intersection with null capabilities", function() {
        var caps = new MessageCapabilities(ALGOS, LANGUAGES, FORMATS);
        var intersectionA = MessageCapabilities.intersection(null, caps);
        var intersectionB = MessageCapabilities.intersection(caps, null);
        
        expect(intersectionA).toBeNull();
        expect(intersectionB).toBeNull();
    });
});

File: tests/src/test/cpp/msg/MessageCapabilitiesTest.cpp
/**
 * Copyright (c) 2016-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <gtest/gtest.h>
#include <io/MslObject.h>
#include <io/MslArray.h>
#include <io/DefaultMslEncoderFactory.h>
#include <io/MslEncoderFormat.h>
#include <msg/MessageCapabilities.h>
#include <MslConstants.h>
#include <MslEncodingException.h>
#include <algorithm>

using namespace std;
using namespace netflix::msl::io;

namespace netflix {
namespace msl {
namespace msg {

namespace {
const std::string KEY_COMPRESSION_ALGOS = "compressionalgos";
const std::string KEY_LANGUAGES = "languages";
const std::string KEY_ENCODER_FORMATS = "encoderformats";

template <typename T> string op(const T& x) { return x.toString(); }

//string toString(MessageCapabilities mc) {
//    MslEncoderFactory mef;
//    const ByteArray ba = mc.toMslEncoding(mef, MslEncoderFormat::JSON);
//    return string(ba.begin(), ba.end());
//}

}

class MessageCapabilitiesTest : public ::testing::Test
{
protected:
    MessageCapabilitiesTest() {
        compressionAlgos_.insert(MslConstants::CompressionAlgorithm::NOCOMPRESSION);
        compressionAlgos_.insert(MslConstants::CompressionAlgorithm::LZW);
        compressionAlgos_.insert(MslConstants::CompressionAlgorithm::GZIP);
        languages_.push_back("aa");
        languages_.push_back("cc");
        languages_.push_back("bb");
        encoderFormats_.insert(MslEncoderFormat::JSON);
    }
    set<MslConstants::CompressionAlgorithm> compressionAlgos_;
    vector<string> languages_;
    set<MslEncoderFormat> encoderFormats_;
};

TEST_F(MessageCapabilitiesTest, simpleCtor)
{
    MessageCapabilities mc(compressionAlgos_, languages_, encoderFormats_);
    EXPECT_EQ(compressionAlgos_, mc.getCompressionAlgorithms());
    EXPECT_EQ(languages_, mc.getLanguages());
    EXPECT_EQ(encoderFormats_, mc.getEncoderFormats());
}

TEST_F(MessageCapabilitiesTest, mslObjectCtor)
{
    // happy path
    shared_ptr<MslObject> mo = make_shared<MslObject>();
    vector<string> ca(compressionAlgos_.size());
    transform(compressionAlgos_.begin(), compressionAlgos_.end(), ca.begin(), op<MslConstants::CompressionAlgorithm>);
    shared_ptr<MslArray> cama = make_shared<MslArray>(ca);
    mo->put(KEY_COMPRESSION_ALGOS, cama);
    shared_ptr<MslArray> lma = make_shared<MslArray>(languages_);
    mo->put(KEY_LANGUAGES, lma);
    vector<string> ef(encoderFormats_.size());
    transform(encoderFormats_.begin(), encoderFormats_.end(), ef.begin(), op<MslEncoderFormat>);
    shared_ptr<MslArray> efma = make_shared<MslArray>(ef);
    mo->put(KEY_ENCODER_FORMATS, efma);
    shared_ptr<MessageCapabilities> mc1 = make_shared<MessageCapabilities>(mo);
    EXPECT_EQ(compressionAlgos_, mc1->getCompressionAlgorithms());
    EXPECT_EQ(languages_, mc1->getLanguages());
    EXPECT_EQ(encoderFormats_, mc1->getEncoderFormats());

    // missing compression algos field
    mo->remove(KEY_COMPRESSION_ALGOS);
    shared_ptr<MessageCapabilities> mc2 = make_shared<MessageCapabilities>(mo);
    EXPECT_TRUE(mc2->getCompressionAlgorithms().empty());
    mo->put(KEY_COMPRESSION_ALGOS, cama);

    // unsupported compression algorithm
    cama->put<string>(-1, "invalidCompressionAlgoName");
    mo->put(KEY_COMPRESSION_ALGOS, cama);
    shared_ptr<MessageCapabilities> mc3 = make_shared<MessageCapabilities>(mo);
    EXPECT_EQ(compressionAlgos_, mc3->getCompressionAlgorithms());

    // missing languages
    mo->remove(KEY_LANGUAGES);
    shared_ptr<MessageCapabilities> mc4 = make_shared<MessageCapabilities>(mo);
    EXPECT_TRUE(mc4->getLanguages().empty());
    mo->put(KEY_LANGUAGES, lma);

    // missing encoder formats
    mo->remove(KEY_ENCODER_FORMATS);
    shared_ptr<MessageCapabilities> mc5 = make_shared<MessageCapabilities>(mo);
    EXPECT_TRUE(mc5->getEncoderFormats().empty());
    mo->put(KEY_ENCODER_FORMATS, efma);

    // unsupported encoder formats
    efma->put<string>(-1, "invalidEncoderFormat");
    mo->put(KEY_ENCODER_FORMATS, efma);
    shared_ptr<MessageCapabilities> mc6 = make_shared<MessageCapabilities>(mo);
    EXPECT_EQ(encoderFormats_, mc6->getEncoderFormats());
}

TEST_F(MessageCapabilitiesTest, toMslEncoding)
{
	// FIXME: This is an error-prone test because there is no guarantee of how
	// a JSON encoding might be constructed or formatted.
    shared_ptr<MessageCapabilities> mc1 = make_shared<MessageCapabilities>(compressionAlgos_, languages_, encoderFormats_);
    shared_ptr<MslEncoderFactory> mef = make_shared<DefaultMslEncoderFactory>();
    shared_ptr<ByteArray> ba1 = mc1->toMslEncoding(mef, MslEncoderFormat::JSON);
    EXPECT_EQ(
        "{\"compressionalgos\":[\"GZIP\",\"LZW\",\"NOCOMPRESSION\"],\"encoderformats\":[\"JSON\"],\"languages\":[\"aa\",\"cc\",\"bb\"]}",
        string(ba1->begin(), ba1->end()));

    shared_ptr<MslObject> mo = make_shared<MslObject>();
    shared_ptr<MessageCapabilities> mc2 = make_shared<MessageCapabilities>(mo);
    shared_ptr<ByteArray> ba2 = mc2->toMslEncoding(mef, MslEncoderFormat::JSON);
    EXPECT_EQ(
        "{\"compressionalgos\":[],\"encoderformats\":[],\"languages\":[]}",
        string(ba2->begin(), ba2->end()));
}

TEST_F(MessageCapabilitiesTest, equality)
{
    set<MslConstants::CompressionAlgorithm> compressionAlgos = compressionAlgos_;
    vector<string> languages = languages_;
    set<MslEncoderFormat> encoderFormats = encoderFormats_;

    shared_ptr<MessageCapabilities> mc1 = make_shared<MessageCapabilities>(compressionAlgos_, languages_, encoderFormats_);
    shared_ptr<MessageCapabilities> mc2 = make_shared<MessageCapabilities>(compressionAlgos, languages, encoderFormats);
    EXPECT_TRUE(*mc1 == *mc2);
    EXPECT_FALSE(*mc1 != *mc2);
    EXPECT_EQ(*mc1, *mc2);

    compressionAlgos.erase(MslConstants::CompressionAlgorithm::LZW);
    shared_ptr<MessageCapabilities> mc3 = make_shared<MessageCapabilities>(compressionAlgos, languages, encoderFormats);
    EXPECT_FALSE(*mc2 == *mc3);
    EXPECT_TRUE(*mc2 != *mc3);
    EXPECT_NE(*mc1, *mc3);
}

TEST_F(MessageCapabilitiesTest, intersection)
{
    // empty intersection with empty should be empty
    const set<MslConstants::CompressionAlgorithm> emptyCompressionAlgos;
    const vector<string> emptyLanguages;
    const set<MslEncoderFormat> emptyEncoderFormats;
    shared_ptr<MessageCapabilities> mc1 = make_shared<MessageCapabilities>(emptyCompressionAlgos, emptyLanguages, emptyEncoderFormats);
    shared_ptr<MessageCapabilities> mc2 = make_shared<MessageCapabilities>(emptyCompressionAlgos, emptyLanguages, emptyEncoderFormats);
    shared_ptr<MessageCapabilities> result = MessageCapabilities::intersection(mc1, mc2);
    EXPECT_TRUE(result->getCompressionAlgorithms().empty());
    EXPECT_TRUE(result->getEncoderFormats().empty());
    EXPECT_TRUE(result->getLanguages().empty());

    // empty intersection with non-empty object should kill the object x3
    shared_ptr<MessageCapabilities> mc3 = make_shared<MessageCapabilities>(compressionAlgos_, languages_, encoderFormats_);
    shared_ptr<MessageCapabilities> emptyMc1 = make_shared<MessageCapabilities>(emptyCompressionAlgos, languages_, encoderFormats_);
    shared_ptr<MessageCapabilities> emptyMc2 = make_shared<MessageCapabilities>(compressionAlgos_, emptyLanguages, encoderFormats_);
    shared_ptr<MessageCapabilities> emptyMc3 = make_shared<MessageCapabilities>(compressionAlgos_, languages_, emptyEncoderFormats);

    result = MessageCapabilities::intersection(mc3, emptyMc1);
    EXPECT_TRUE(result->getCompressionAlgorithms().empty());
    EXPECT_NE(compressionAlgos_, result->getCompressionAlgorithms());
    EXPECT_EQ(languages_, result->getLanguages());
    EXPECT_EQ(encoderFormats_, result->getEncoderFormats());

    result = MessageCapabilities::intersection(mc3, emptyMc2);
    EXPECT_TRUE(result->getLanguages().empty());
    EXPECT_EQ(compressionAlgos_, result->getCompressionAlgorithms());
    EXPECT_NE(languages_, result->getLanguages());
    EXPECT_EQ(encoderFormats_, result->getEncoderFormats());

    result = MessageCapabilities::intersection(mc3, emptyMc3);
    EXPECT_TRUE(result->getEncoderFormats().empty());
    EXPECT_EQ(compressionAlgos_, result->getCompressionAlgorithms());
    EXPECT_EQ(languages_, result->getLanguages());
    EXPECT_NE(encoderFormats_, result->getEncoderFormats());

    // intersection between two non-empty objects should be the lesser object
    set<MslConstants::CompressionAlgorithm> compressionAlgos = compressionAlgos_;
    vector<string> languages = languages_;
    set<MslEncoderFormat> encoderFormats = encoderFormats_;
    compressionAlgos.erase(MslConstants::CompressionAlgorithm::LZW);
    languages.erase(remove(languages.begin(), languages.end(), "bb"), languages.end());
    encoderFormats.erase(MslEncoderFormat::JSON);
    shared_ptr<MessageCapabilities> mc4 = make_shared<MessageCapabilities>(compressionAlgos, languages, encoderFormats);
    result = MessageCapabilities::intersection(mc3, mc4);
    EXPECT_EQ(compressionAlgos, result->getCompressionAlgorithms());
    EXPECT_EQ(languages, result->getLanguages());
    EXPECT_EQ(encoderFormats, result->getEncoderFormats());

    // intersection order should not matter
    shared_ptr<MessageCapabilities> result1 = MessageCapabilities::intersection(mc4, mc3);
    EXPECT_EQ(*result, *result1);
}

}}} // namespace netflix::msl::msg
